shared/src/main/scala/mlscript/TypingUnit.scala

```diff
@@ -0,0 +1,155 @@
+    // `class A` ==> `class A {}`
+    case Top => Nil -> N
+    // `class A { <items> }` ==>
+    // `class A { constructor(fields) { <items> } }`
+    case Record(fields) => fields.map(_._1.name) -> N
+    // `class B: A` ==> `class B extends A {}`
+    // If `A` is a type alias, it is replaced by its real type.
+    // Otherwise just use the name.
+    case TypeName(name) =>
+      if (traitNames contains name)
+        Nil -> N
+      else
+        typeAliasMap get name match {
+          // The base class is not a type alias.
+          case N => Nil -> S(name)
+          // The base class is a type alias with no parameters.
+          // Good, just make sure all term is normalized.
+          case S(Nil -> body) => getBaseClassAndFields(substitute(body))
+          // The base class is a type alias with parameters.
+          // Oops, we don't support this.
+          case S(tparams -> _) =>
+            throw TypingUnitError(
+              s"type $name expects ${tparams.length} type parameters but nothing provided"
+            )
+        }
+    // `class C: { <items> } & A` ==>
+    // `class C extends A { constructor(fields) { <items> } }`
+    case Inter(Record(entries), ty) =>
+      val (fields, cls) = getBaseClassAndFields(ty)
+      entries.map(_._1.name) ++ fields -> cls
+    // `class C: { <items> } & A` ==>
+    // `class C extends A { constructor(fields) { <items> } }`
+    case Inter(ty, Record(entries)) =>
+      val (fields, cls) = getBaseClassAndFields(ty)
+      fields ++ entries.map(_._1.name) -> cls
+    // `class C: <X> & <Y>`: resolve X and Y respectively.
+    case Inter(ty1, ty2) =>
+      val (fields1, cls1) = getBaseClassAndFields(ty1)
+      val (fields2, cls2) = getBaseClassAndFields(ty2)
+      (cls1, cls2) match {
+        case (N, N) =>
+          fields1 ++ fields2 -> N
+        case (N, S(cls)) =>
+          fields1 ++ fields2 -> S(cls)
+        case (S(cls), N) =>
+          fields1 ++ fields2 -> S(cls)
+        case (S(cls1), S(cls2)) =>
+          if (cls1 === cls2) {
+            fields1 ++ fields2 -> S(cls1)
+          } else {
+            throw TypingUnitError(s"Cannot have two base classes: $cls1, $cls2")
+          }
+      }
+    // `class C: F[X]` and (`F[X]` => `A`) ==> `class C extends A {}`
+    // For applied types such as `Id[T]`, normalize them before translation.
+    // Do not forget to normalize type arguments first.
+    case AppliedType(TypeName(base), targs) =>
+      if (traitNames contains base)
+        Nil -> N
+      else
+        getBaseClassAndFields(applyTypeAlias(base, targs map { substitute(_) }))
+    // There is some other possibilities such as `class Fun[A]: A -> A`.
+    // But it is not achievable in JavaScript.
+    case Rem(_, _) | TypeVar(_, _) | Literal(_) | Recursive(_, _) | Bot | Top |
+        Tuple(_) | Neg(_) | Bounds(_, _) | WithExtension(_, _) |
+        Function(_, _) | Union(_, _) | Arr(_) =>
+      throw TypingUnitError(s"unable to derive from type $ty")
```

Comment:
I know this is the error message @chengluyu wrote, but it's very poor and should be improved. What "cannot derive"?